Fetch data at build time

jokroese opened the pull request:
Instead of sending the main data (ids and coordinates) to each user directly from Supabase, this PR gets the data once at build time.

This PR changes our data fetching strategy by obtaining the main data (ids and coordinates) from Supabase once at build time, rather than querying the database for each user. The fetched data is stored as a json file in the static directory.

This has some benefits:

- **Improved developer experience**: Developers don't need to set up Supabase locally (which is a bit of a pain in the ass), which simplifies onboarding. Instead, they can simply place a json file in the static directory, making it easier to get started.
- **Performance: reduce queries**: by fetching the data once during the build process, we remove multiple database queries, reducing load and potential latency. we don't need to do a database query for each user and instead just send the pre-rendered json
- **Performance: data delivery** 2: Pre-rendered json files are served directly to users, improving response times and overall user experience.
- **Setting up for the future**: this is also the route we'd go down if we want to use PMTiles for the moments data

Considerations:

- The main drawback is that the map will not always display the most recent data. However, talking to @queeringthemap, this isn't a worry. As adding data depends on moderators approving them, it is fine to have regular deployments (through GitHub Actions or similar). If we need to change data quicker than that – removing a problematic moment – we can just do a manual deploy.

narcode reviewed (commented):
I think the code is right and it is true that because new moments need to be approved first we don't need realtime data. However I want us to think about the following:

- Is this scalable?
How far do we think we can scale this, for example with the total amount of moments in 1 year? We could maybe first test locally (or add a test) with a very very extreme large moments.json and see if its solving the problems we intend to solve with this PR. 

- When/how to do new deploys?
Maybe trigger a new deploy automatically everytime a moment gets approved.

- Personal thoughts on the benefits
**Improved developer experience** ✅  agreed.
**Performance:** reduce queries ✅ agreed but scalling the concurrent data fetching on the API/DB level should not be difficult. It generally means scalling the infrastructure when needed and/or adding a cache layer.

Performance: data delivery: ❓ This one I am not so sure about because users with bad internet connections will also have trouble downloading the `moments.json` file even if it is static, at least the first time they visit the site. I still feel the most efficient data delivery strategy will be to query the DB for the points within the user's viewport. and re-fetch on zoom events. Of find a way to "stream" the points and render them as they arrive. I do like however that this is already some sort of caching strategy which does make a lot of sense.

Setting up for the future: ✅

jokroese commented on src/lib/scripts/fetchData.ts:
```diff
@@ -0,0 +1,46 @@
+	const geoJson = await fetchData();
+
+	if (!geoJson) {
+		console.error('Failed to fetch data, aborting file write.');
```
If we don't have the data the build should probably abort though, otherwise we could deploy an empty map.

narcode replied:
ok, let's maybe write a test to make sure the build aborts (could be a github action test trying to build)?

jokroese commented:
Great reflections. Underlying all of this topic is that I think the future of delivering the points data is using PMTiles. So everything specific to json is probably a stopping point on the way to that.

> * Is this scalable?

I think it's pretty scalable but if we move to PMTiles, this doesn't need to scale too far.

> * When/how to do new deploys?
> Maybe trigger a new deploy automatically everytime a moment gets approved.

Triggering on each moment will be a bit wild because a moderator can approve ~10 moments a minute! We can check every hour if there's been any moderation, and if there is, we can trigger a deploy. Even every day would be totally fine. 

> **Performance:** reduce queries ✅ agreed but scalling the concurrent data fetching on the API/DB level should not be difficult. It generally means scalling the infrastructure when needed and/or adding a cache layer.

For sure. Scaling with the API/DB is definitely very doable. Just wondering if it's necessary.

> Performance: data delivery: ❓ This one I am not so sure about because users with bad internet connections will also have trouble downloading the `moments.json` file even if it is static, at least the first time they visit the site. I still feel the most efficient data delivery strategy will be to query the DB for the points within the user's viewport. and re-fetch on zoom events. Of find a way to "stream" the points and render them as they arrive. I do like however that this is already some sort of caching strategy which does make a lot of sense.

This idea is built into PMTiles and I think it will be much more efficient to use that implementation rather than building our own. In the meantime, grouping the moments together means they compress further than if they are separate, so it is less data in total that is sent over. Then we get caching (the PMTiles) and partial delivery in one.

jokroese commented:
The options for next steps:

1. We merge this for now and work to improve it
1. We reject it for now
1. We do some stress testing of lots of data
1. We implement PMTiles already in this branch

What should we do with this for now?

narcode commented:
If PMTiles **is not** within the 1.0 scope I suggest:

1 . We quickly stress test with lots of data -> if successful we merge (i don't expect real problems actually)
2. we make sure that the build will fail if the fetching fails

if PMTiles **is** within 1.0 scope:
- we implement PMTiles already

In the end making this change should not be a blocker so let's go 🚀 :)

jokroese commented:
As @narcode suggested,

- I stress-tested this approach, using 999,999 points which is way bigger than we'll have for a while (we currently have ~150,000). All seems fine. With this many points, the compressed data is 12MB which is big of course but is still <50% the size of what is sent in the current deployment.
- I implemented the build failing if the fetch fails

I also:

- wrote a seed script for creating a local json. This is useful for easy stress testing and also for quick setup of the project.
- documented the quick setup approach in the README
- added shortcuts for fetching data and seeding data, independent of developing and building the project
